omnigibson/simulator.py

```diff
@@ -1322,43 +1407,82 @@ def get_metadata(self, key):
+                states.append(state)
 
-                # Override the init info with our json path
-                init_info["args"]["scene_file"] = json_path
+                if load_from_scratch:
+                    # Override the init info with our json path
+                    init_info["args"]["scene_file"] = scene_file
 
-                # Also make sure we have any additional modifications necessary from the specific scene
-                og.REGISTERED_SCENES[init_info["class_name"]].modify_init_info_for_restoring(init_info=init_info)
+                    # Also make sure we have any additional modifications necessary from the specific scene
+                    og.REGISTERED_SCENES[init_info["class_name"]].modify_init_info_for_restoring(init_info=init_info)
 
-                # Recreate and import the saved scene
-                og.sim.stop()
-                recreated_scene = create_object_from_init_info(init_info)
-                self.import_scene(scene=recreated_scene)
+                    # Recreate and import the saved scene
+                    recreated_scene = create_object_from_init_info(init_info)
+                    self.import_scene(scene=recreated_scene)
 
-                # Start the simulation and restore the dynamic state of the scene and then pause again
-                self.play()
-                recreated_scene.load_state(state, serialized=False)
+                else:
+                    scene = self.scenes[i]
+                    # Make sure the class type is the same
+                    if scene.__class__.__name__ != init_info["class_name"]:
+                        log.error(
+                            f"Got mismatch in scene type: current is type {scene.__class__.__name__}, trying to load type {init_info['class_name']}"
+                        )
+
+                    current_obj_names = set(scene.object_registry.get_dict("name").keys())
+                    load_obj_names = set(scene_info["objects_info"]["init_info"].keys())
+
+                    objs_to_remove = current_obj_names - load_obj_names
+                    objs_to_add = load_obj_names - current_obj_names
+
+                    # Delete any extra objects that currently exist in the scene stage
+                    for obj_to_remove in objs_to_remove:
+                        obj = scene.object_registry("name", obj_to_remove)
+                        og.sim.remove_object(obj)
+
+                    # Add any extra objects that do not currently exist in the scene stage
+                    for obj_to_add in objs_to_add:
+                        obj = create_object_from_init_info(scene_info["objects_info"]["init_info"][obj_to_add])
+                        scene.add_object(obj)
+
+            # Start the simulation and restore the dynamic state of the scene and then pause again
+            self.play()
+            for i, state in enumerate(states):
+                self.scenes[i].load_state(state, serialized=False)
```

Comment:
wow very clean! Where is the `pause` tho?

Reply:
removed outdated comment